Remove repeated random string generations in scheduler volume predicate

shyamjvs opened the pull request:
Ref https://github.com/kubernetes/kubernetes/issues/53327

@wojtek-t @liggitt @jsafrane - Does this look ok to you?

shyamjvs committed:
Remove repeated random string generations in scheduler volume predicate

shyamjvs commented:
I'll try running kubemark-5000 with this change and see the outcome.

liggitt commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -196,15 +205,25 @@ type VolumeFilter struct {
 
+func (c *MaxPDVolumeCountChecker) getNextVolumeIDPrefix() string {
+	c.lock.Lock()
+	defer c.lock.Unlock()
```
can be lock-free with `atomic.AddInt64(&c.volumeIDPrefixCounter, 1)`

shyamjvs replied:
Smooth! Changed it.

shyamjvs commented:
@liggitt Fixed stuff. PTAL.

liggitt commented:
change LGTM, will let sig-storage weigh in

shyamjvs commented:
cc @kubernetes/sig-storage-pr-reviews

msau42 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -266,10 +282,13 @@ func (c *MaxPDVolumeCountChecker) predicate(pod *v1.Pod, meta algorithm.Predicat
+	// predicate() call so a deleted PVC used by two pods is counted as one volume
+	// and not as two. Also, appending a counter to a const string removes
+	// the need to create a random string each time (which is expensive).
+	randomPrefix := c.getNextVolumeIDPrefix()
```
I'm not sure if a random prefix is needed for each predicate call.  It may be good enough to just generate a random prefix once during initialization.  All the structures used to count volumes in this predicate are local and shouldn't conflict if the predicate is running in parallel.

shyamjvs replied:
That's precisely what I'm doing in this PR - a random prefix initialized once and then just appending it to a counter. Do you mean sth else?

msau42 replied:
I mean, I don't think you even need a counter.  Just a random prefix initialized in the beginning.

shyamjvs replied:
In that case do we even need the initialized string to be random? For e.g why not just set it to some hard-coded value like 'foobar'?

liggitt replied:
don't you risk conflicting with actual object names?

msau42 replied:
Yes, we just want a prefix that won't conflict with actual object names.  So we could generate some random prefix, or have some hardcoded value that we are sure nobody would ever use.  I don't think it needs to be generated every time we run the predicate though.

shyamjvs replied:
Changed it - PTAL.

shyamjvs commented:
@kubernetes/test-infra-maintainers Seems like there's a bug with the `do-not-merge/release-note-label-needed` label. Is it intended to add it (and remove the release-note-none label) each time I change the PR title? Also, even if I added the `release-note-none` label during PR creation, it still removed it and added the do-not-merge one. Other e.g.s - https://github.com/kubernetes/kubernetes/pull/54045, https://github.com/kubernetes/kubernetes/pull/53977. This only started happening recently.

msau42 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -196,16 +201,17 @@ type VolumeFilter struct {
+		maxVolumes:           maxVolumes,
+		pvInfo:               pvInfo,
+		pvcInfo:              pvcInfo,
+		randomVolumeIDPrefix: rand.String(8),
```
Can this be 32 like originally?  That way we reduce the chance of conflicting with the user's object names.

shyamjvs replied:
Done.

cjwagner commented:
@shyamjvs The release note process is WAI.
If you want to specify that there is no release note you should either:
- make the release note "none" via the `release-note` block in the PR body like this: ```` ```release-note none ``` ````
- or use the `/release-note-none` command

/release-note-none

shyamjvs commented:
@cjwagner Sounds good. But any idea why the label is getting removed on renaming the PR title?

msau42 commented:
/lgtm

cjwagner commented:
@shyamjvs Yeah, the PR 'edit' event is used to detect when the PR body text (including the release-note block) is changed, but it is also triggered by title changes. The Prow plugin that handled the event saw the `release-note-none` label on the PR and assumed that it was previously added to the PR based on the release-note block being "none" since there was no `/release-note-none` command on the PR. Since the plugin didn't find a release-note block after the edit event, it thinks that you removed the release-note block which is why it applies the `release-note-needed` label.

shyamjvs commented:
@davidopp / @bsalamat Could one of you please approve the PR?

bsalamat commented:
/approve

shyamjvs commented:
/test pull-kubernetes-e2e-gce
